src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -312,23 +296,11 @@ private void SetStartHelper(Delegate start, int maxStackSize)
         private extern void StartupSetApartmentStateInternal();
 #endif // FEATURE_COMINTEROP_APARTMENT_SUPPORT
 
-        /// <summary>Retrieves the name of the thread.</summary>
-        public string Name
+        partial void ThreadNameChanged(string value)
```

Comment:
This still does not prevent the managed and unmanaged sides name from getting out of sync. Consider one thread setting the name to `A` and second thread setting the name to `B`. You can end up with the name on the managed side set to `A` and the name on the unmanaged side set to `B`.

I think that the setting of the `_name` field has to stay under the lock.

Reply:
I don't think you can. The Interlocked construct in `Name` setter will allow only one thread to get as far as updating the name on unmanaged side.

Reply:
@jkotas I think I see where you are heading but shouldn't the unmanaged code lock on `this` as well (there is no explicit locking right now).

Reply:
> I don't think you can

Consider this:
```
Thread 1: _name = "A";
Thread 1: ThreadNameChanged("A")
                                          Thread 2: _name = "B";
                                          Thread 2: ThreadNameChanged("B")
                                                          lock { InformThreadNameChange("B"); }          
Thread 2: lock { InformThreadNameChange("A") }
```
Result: Managed side thinks that the name is "B", unmanaged side thinks that the name is "A"

> shouldn't the unmanaged code lock on `this`

I do not see why this would help.

This API is not performance critical. There is no point in inventing lock-free code here.

Reply:
Also, I do not think this `InvalidOperation_WriteOnce` block is useful. I do not see a problem with changing the thread name multiple times - but that can be done as a separate fix.

Reply:
@jkotas I see the possibility of your scenario with the old code, but I still fail to see it with the new code for `Name`.

`Thread 2: _name = "A";` can no longer happen because the access to `_name` is Interlocked and it will throw on Thread B.

Reply:
> access to `_name` is Interlocked and it will throw on Thread B.

The throw condition does not work if other thread sets it to null. (I did not have my example written with null originally, but edited it to fix it.)

Reply:
I see :)